Yzhou2001 gcs 106

yzhou2001 opened the pull request.

medb commented:
Please, use https://github.com/google/google-java-format to format your change

yzhou2001 commented:
Codes reformatted per request. Note that there are many formatting changes on my changed files but not resulting from my changes, which I just leave as they are changed.

The CLA error will need some time to clear up  though. Thanks.

medb commented:
@yzhou2001 please, format only modified lines instead of whole files:
https://github.com/google/google-java-format/blob/master/scripts/google-java-format-diff.py

yzhou2001 commented:
I tried google-java-format-diff.py with the sample usage:
%git diff -U0 HEAD^ | ~/bin/google-java-format-diff.py -p1 -i

(or, a slight variation:)
git diff -U0 75fa3841b4c0a0a1b8520d64e2240a7709dff75a 8abcd0ddc76139ecf01af7159ad1598ca9602a3b|~/bin/google-java-format-diff.py -p1 -i

but got exception as below:
Traceback (most recent call last):
  File "/Users/yazhou/bin/google-java-format-diff.py", line 118, in <module>
    main()
  File "/Users/yazhou/bin/google-java-format-diff.py", line 101, in main
    stderr=None, stdin=subprocess.PIPE)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/usr/local/Cellar/python@2/2.7.15_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1047, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory

Did I use the tool correctly ? My Python version is Python 2.7.15. Or maybe you can send me your command line of using the tool.

It's my first time to use Google's java formatters. Please bear with me with simple questions. Thanks.

medb commented:
It worked for me with next command:
```
git diff -U0 HEAD~2 | python google-java-format-diff.py -p1 -i
```

You also may want to allow me to edit this PR, so I can help with fixing it too:
https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

yzhou2001 commented:
It seems that the permission has already been given:

<img width="1223" alt="screen shot 2018-09-06 at 10 25 28 am" src="https://user-images.githubusercontent.com/8181800/45174365-869b6180-b1bf-11e8-8895-92dcbc97990f.png">

medb commented:
I have it in `/usr/bin/`
```bash
$ ls -al /usr/bin/google-java-format
lrwxrwxrwx 1 root root 49 Dec 31  1999 /usr/bin/google-java-format -> /usr/lib/google-java-format/google-java-format.sh
```

I think you can create `google-java-format.sh` with next content and make it executable:
```bash
java -jar /path/to/google-java-format-1.6-all-deps.jar "$@"
```

yzhou2001 commented:
Changes checked in. I used a slightly different approach as suggested in https://github.com/google/google-java-format/pull/227/files

Anyhow, thanks for help.

yzhou2001 commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:
```diff
@@ -1197,29 +1196,47 @@ private void onCopyFailure(
-   * @param listedPrefixes output parameter into which retrieved prefixes will be added
+   * @param pageState if pagination enabled, holds stateful info across pages
    */
   private void listStorageObjectsAndPrefixes(
```
The "page state" carried across different calls does not only have the pageToken. The "prefixes" needs to be kept in order to keep the original processing logic unchanged. For "listedObjects", it's reused across calls to relieve some GC pressure.

yzhou2001 commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:
```diff
@@ -1353,38 +1368,53 @@ private static long getMaxRemainingResults(
-    for (StorageObject obj : listedObjects) {
+    List<String> objectNames = new ArrayList<>(pageState.getPrefixes().size());
+    objectNames.addAll(pageState.getPrefixes());
+    ;
```
will do.

yzhou2001 commented on gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java:
```diff
@@ -1031,15 +1031,15 @@ private FileInfo repairPossibleImplicitDirectory(FileInfo pathInfo)
    * @param prefix the prefix to use to list all matching objects.
    */
-  public List<FileInfo> listAllFileInfoForPrefix(URI prefix)
+  public List<FileInfo> listAllFileInfoForPrefix(URI prefix, PageState pageState)
```
I think we need to differentiate "public methods" from "public APIs". For latter the backward compatibility had better to be maintained carefully; for the former it's not a requirement. In particular, this class of GoogleCloudStorageFileSystem, as it seems to me, serves as a bridge between the GCS implementation of Hadoop FileSystem, and the GoogleCloudStorage which handles the GCS calls. So this class is tightly coupled with the two and should not be used standalone beyond testing. So it's hard to think this class, and its methods, should be exposed to users as a public API.

medb replied:
Some users that need high-level GCS abstraction, but do not need HDFS-compatibility are using `GoogleCloudStorageFileSystem` class directly, especially outside Hadoop ecosystem

yzhou2001 replied:
ok. will change

yzhou2001 commented:
The fix of GoogleCloudStorageFileSystem is pushed in.

I just noticed the fix won't work for Hadoop 1. Is Hadoop 1 still supported?

medb commented:
Yes, we still need to be compatible with Hadoop 1

medb commented:
@yzhou2001 I have simplified an implementation and fixed few bugs and tests, but I can not revert `InMemoryGlobberFileSystem` from GitHub UI. May you revert it?

yzhou2001 commented:
> @yzhou2001 I have simplified an implementation and fixed few bugs and tests, but I can not revert `InMemoryGlobberFileSystem` from GitHub UI. May you revert it?

I tried with the addition of the missing file, but encountered one test case failure that caused interruption of the automated test process.

A further scrutiny reveals that your changes are not just a simplification: it actually changes the processing logic, which is what I strived to keep intact at first place. Basically you changed a stateful page-wise call to a stateless one. Please check one of my previous comments on this for more details.

As for use of the InMemoryGlobberFileSystem class, note that the fundamental problem of the original problem is that the processing turned an ought-to-be nonblocking processing to a blocking processing, resulting in huge memory waste. This version of your fix improves from a total blocking call to a partial, page-wise blocking call; while my version is a nonblocking version and should be more efficient in general.

As for the "few bugs", could you elaborate more ? Thanks.

medb commented:
Yes, there will be 1 test failure that I have a fix for, as soon as you will revert this file, I will apply it.

There were failures in integration tests, because `toFileStatusesWithImplicitDirectories` is called after globbing and it was adding unnecessary implicit directories.

I would prefer to use `InMemoryGlobberFileSystem` because [globbing logic](https://github.com/apache/hadoop/blob/184544eff8b3d5951e4f04d080fa00de8e1eec95/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Globber.java#L155) is actually more complicated then just filtering out non-matched patterns.

What value brings statefulness? Looks like it only complicates implementation without significant benefits.

medb commented:
To fix failing test after reverting `InMemoryGlobberFileSystem` class you will need just delete [this statement](https://github.com/GoogleCloudPlatform/bigdata-interop/blob/4a440707bd9fb7af2ce4d49e9ecab55992105254/gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InMemoryGlobberFileSystem.java#L57:L59).

May you run your globbing tests and see how memory consumption compares with stateful vs stateless pagination?

yzhou2001 commented:
> To fix failing test after reverting `InMemoryGlobberFileSystem` class you will need just delete [this statement](https://github.com/GoogleCloudPlatform/bigdata-interop/blob/4a440707bd9fb7af2ce4d49e9ecab55992105254/gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InMemoryGlobberFileSystem.java#L57:L59).
> 
> May you run your globbing tests and see how memory consumption compares with stateful vs stateless pagination?

Just add the file back. I still suggest not to use the InMemoryFileSystem in list, and use the nonblocking approach of mine. A processing principle is never turn a nonblocking processing into a blocking one. If InMemoryGlobberFileSystem was used in other places, it should be in. As for the "more" complex doGlob() of Hadoop, yes. But the purpose is to select from an existing FS; while we are building a FS first before select second. This is a total waste. We do not need to build at all: just build the end result while filtering in a streaming fashion!

yzhou2001 commented:
> Yes, there will be 1 test failure that I have a fix for, as soon as you will revert this file, I will apply it.
> 
> There were failures in integration tests, because `toFileStatusesWithImplicitDirectories` is called after globbing and it was adding unnecessary implicit directories.
> 
> I would prefer to use `InMemoryGlobberFileSystem` because [globbing logic](https://github.com/apache/hadoop/blob/184544eff8b3d5951e4f04d080fa00de8e1eec95/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Globber.java#L155) is actually more complicated then just filtering out non-matched patterns.
> 
> What value brings statefulness? Looks like it only complicates implementation without significant benefits.

The stateful page processing was originally and I think it makes sense. I kept it; while in your version, one call path keeps it, the other does not!

yzhou2001 commented:
As for the bug you encountered, I ran and past all tests before created the PR. If this turns out to be a bug, have added a test case to cover it?

medb commented:
> As for the bug you encountered, I ran and past all tests before created the PR. If this turns out to be a bug, have added a test case to cover it?

Did you run integration test? By default `mvn -Phadoop2 clean test` will run only unit tests.

medb commented:
Regarding `InMemoryGlobberFileSystem`, we can delete it when all tests will be passing, but better to do this in separate PR, because it not strictly related to this change.

yzhou2001 commented:
> > As for the bug you encountered, I ran and past all tests before created the PR. If this turns out to be a bug, have added a test case to cover it?
> 
> Did you run integration test? By default `mvn -Phadoop2 clean test` will run only unit tests.

No. My fault.

yzhou2001 commented:
> Regarding `InMemoryGlobberFileSystem`, we can delete it when all tests will be passing, but better to do this in separate PR, because it not strictly related to this change.

Do you mean without using it there is a test case failure?

medb commented:
To make decision on stateless vs statefulness we need some data to base it on. Let's make it when you will have memory consumption data.

yzhou2001 commented:
> > Regarding `InMemoryGlobberFileSystem`, we can delete it when all tests will be passing, but better to do this in separate PR, because it not strictly related to this change.
> 
> Do you mean without using it there is a test case failure?

If so, there must be somewhere else wrong. Note that Hadoop glob has been used and tested over years, and I don't expect naiive bugs there.

medb commented:
> Do you mean without using it there is a test case failure?

Yes, `InMemoryGlobberFileSystem` has some logic that filter outs implicit directories added in `toFileStatusesWithImplicitDirectories` method.

yzhou2001 commented:
> To make decision on stateless vs statefulness we need some data to base it on. Let's make it when you will have memory consumption data.

The concern regarding stateless or stateful is not performance, but correctness.

medb commented:
> The concern regarding stateless or stateful is not performance, but correctness.

What do you mean? They will return different results?

yzhou2001 commented:
> > The concern regarding stateless or stateful is not performance, but correctness.
> 
> What do you mean? They will return different results?

It might, even though our test cases do not cover it yet. The original logic was correct I believe.

yzhou2001 commented:
> > Do you mean without using it there is a test case failure?
> 
> Yes, `InMemoryGlobberFileSystem` has some logic that filter outs implicit directories added in `toFileStatusesWithImplicitDirectories` method.

Do you mean by this code snippet?

for (FileStatus fileStatus : fileStatuses) {
      this.fileStatusesByPath.put(fileStatus.getPath(), fileStatus);
      if (fileStatus.getPath().getParent() != null) {
        this.fileStatusesByParentPath
            .computeIfAbsent(fileStatus.getPath().getParent(), k -> new ArrayList<>())
            .add(fileStatus);
      }
    }

If so, the extra filtering should have happened before putting it into InMemoryFileSystem.

Fundamentally, if the entries in InMemoryFileSystem were *ever* related among themselves, the paginated processing would be impossible, and the processing would have to be a blocking one; if not, they should be processed in a nonblocking and streaming manner.

medb commented:
> Do you mean by this code snippet?

Possibly it handled here.

> if not, they should be processed in a nonblocking and streaming manner.

What do you mean under non-blocking? `InMemoryGlobberFileSystem` globs separately each page without waiting on other pages.

yzhou2001 commented:
> > Do you mean by this code snippet?
> 
> Possibly it handled here.
> 
> > if not, they should be processed in a nonblocking and streaming manner.
> 
> What do you mean under non-blocking? `InMemoryGlobberFileSystem` globs separately each page without waiting on other pages.

Non-blocking by rows which is the logical processing unit unless they are logically related, not by pages, which in my opinion is just a physical partition and of no logical sense: that is, if the rows were related, no paged processing would be possible; if not, row-based processing is preferred while page-based grouping will also work but could incur extra resources.

medb commented:
I see, thanks for explanation. I agree that it would be nicer to filter row-by-row without `InMemoryGlobberFileSystem` but let's do it in separate PR.

Did you have a chance to check memory consumption during globbing with page-by-page filtering?

yzhou2001 commented:
> I see, thanks for explanation. I agree that it would be nicer to filter row-by-row without `InMemoryGlobberFileSystem` but let's do it in separate PR.
> 
> Did you have a chance to check memory consumption during globbing with page-by-page filtering?

No. As said, it could be small, depending the lengths of path. But from the perspective of logic cleanness, easy understanding, ..., etc, we should use a streaming one. In fact, that was the first thing catching my eyes when I took a first look at the issue. From full blocking to page blocking, it'd be a half solution.

In the meanwhile, I believe we need to keep the stateful processing, which is more important.

BTW, how do you run the "integration test" ? Thanks.

medb commented:
You can run integration test with [this](https://github.com/GoogleCloudPlatform/bigdata-interop/blob/master/tools/run_integration_tests.sh) script.

> In the meanwhile, I believe we need to keep the stateful processing, which is more important.

Sure, may you show at least one example where stateless processing will give incorrect result but stateful will give correct one?

yzhou2001 commented:
> You can run integration test with [this](https://github.com/GoogleCloudPlatform/bigdata-interop/blob/master/tools/run_integration_tests.sh) script.
> 
> > In the meanwhile, I believe we need to keep the stateful processing, which is more important.
> 
> Sure, may you show at least one example where stateless processing will give incorrect result but stateful will give correct one?

An imaginary one:

Page 1:

prefix1
path1
path2
-------------------
Page2:

path3
path4

The point through browsing the original code is that the "prefix1" will have some impact on processing rows of "path1", "path2", "path3", and more.  So specifically the leftover prefixes are the "state" across different pages, which we should keep. In your version, the non-paging call path keeps it; the paging call path does not, which is inconsistent at least. It'd be harder to reproduce. You can check with GCP people to see if what I am claiming here is accurate. There is a possibility that the original code logic is already more conservative than necessary and there is already some *logical* considerations when breaking up the data chunk into pages from the GCS server, then you'll be right here. But I am tending to think otherwise now.

yzhou2001 commented:
Do you have a command line example  to run tools/run_integration_tests.sh? Thanks!

medb commented:
```bash
tools/run_integration_tests.sh "hadoop2" "<project_id>" "<service_account_email>" "<path_to_service_file_p12>"
```

medb commented:
I think that case that you have is handled at [de-duplication phase](https://github.com/GoogleCloudPlatform/bigdata-interop/blob/27cf4d0673c07b139fbe71e76cabdf6c4e93c9c3/gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java#L1147:L1156).

yzhou2001 commented:
> I think that case that you have is handled at [de-duplication phase](https://github.com/GoogleCloudPlatform/bigdata-interop/blob/27cf4d0673c07b139fbe71e76cabdf6c4e93c9c3/gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java#L1147:L1156).

By just the GCS connector codes, they don't appear to be logically equivalent. The leftover prefixes are not accumulated and used for next pages.

yzhou2001 commented:
> `prefixes` returned by GCS are only used to create implicit directories and de-duplication logic will handle this if there will be any duplicates between pages.
> 
> Also, from my brief GCS API testing, it looks like GCS never returns duplicate prefixes and objects, even when with pagination.

There might be corner cases to fail that will be difficult to reproduce. My point is that the processing logic has changed. From my limited experience with GCS, I can't tell whether it's perfect or not. I checked the online docs and found no clue. So my best bet is keep the logic intact. You may want to seek comments from more experienced people in this part.
